src/python/pants/option/config.py

```diff
@@ -137,39 +145,93 @@ def _getinstance(self, section, option, type_, default=None):
     return parse_expression(name=key, val=raw_value, acceptable_types=type_,
                             raise_type=self.ConfigError)
 
-  # Subclasses must implement.
+  @abstractmethod
   def configs(self) -> Sequence["_SingleFileConfig"]:
     """Returns the underlying single-file configs represented by this object."""
-    raise NotImplementedError()
 
+  @abstractmethod
   def sources(self) -> List[str]:
     """Returns the sources of this config as a list of filenames."""
-    raise NotImplementedError()
 
+  @abstractmethod
   def sections(self) -> List[str]:
     """Returns the sections in this config (not including DEFAULT)."""
-    raise NotImplementedError()
 
+  @abstractmethod
   def has_section(self, section: str) -> bool:
     """Returns whether this config has the section."""
-    raise NotImplementedError()
 
+  @abstractmethod
   def has_option(self, section: str, option: str) -> bool:
-    """Returns whether this config specified a value the option."""
-    raise NotImplementedError()
+    """Returns whether this config specified a value for the option."""
 
+  @abstractmethod
   def get_value(self, section: str, option: str) -> Optional[str]:
     """Returns the value of the option in this config as a string, or None if no value specified."""
-    raise NotImplementedError()
 
+  @abstractmethod
   def get_source_for_option(self, section: str, option: str) -> Optional[str]:
     """Returns the path to the source file the given option was defined in.
 
     :param section: the scope of the option.
     :param option: the name of the option.
     :returns: the path to the config file, or None if the option was not defined by a config file.
     """
-    raise NotImplementedError
+
+
+class _ConfigValues(ABC):
+  """Encapsulates resolving the actual config values specified by the user's config file.
+
+  Beyond providing better encapsulation, this allows us to support alternative config file formats
+  in the future if we ever decide to support formats other than INI.
+  """
+
+  @abstractmethod
+  def sections(self) -> List[str]:
+    """Returns the sections in this config (not including DEFAULT)."""
+
+  @abstractmethod
+  def has_section(self, section: str) -> bool:
+    pass
+
+  @abstractmethod
+  def has_option(self, section: str, option: str) -> bool:
+    pass
+
+  @abstractmethod
+  def get_value(self, section: str, option: str) -> Optional[str]:
+    pass
+
+  @abstractmethod
+  def options(self, section: str) -> List[str]:
+    """All options defined for the section."""
+
+  @abstractmethod
+  def defaults(self) -> Mapping[str, Any]:
+    """All the DEFAULT values."""
+
+
+@dataclass(frozen=True)
+class _IniValues(_ConfigValues):
```

Comment:
To add TOML support, all that we now need to do is create a `_TomlValues` class and teach `Config._metaload()` when to create an `_IniValues` vs. `_TomlValues` instance (determined by the file suffix).

Same with other file formats like YAML, if the TOML experiment fails and we want to explore alternatives. Simply create a `_YamlValues` class.